rskj-core/src/main/java/co/rsk/net/SnapshotProcessor.java

```diff
@@ -248,20 +451,36 @@ void processSnapBlocksRequestInternal(Peer sender, SnapBlocksRequestMessage requ
     }
 
     public void processSnapBlocksResponse(SnapSyncState state, Peer sender, SnapBlocksResponseMessage responseMessage) {
+        if (!state.isRunning()) {
+            return;
+        }
+
         long lastRequiredBlock = state.getLastBlock().getNumber() - BLOCKS_REQUIRED;
         List<Block> blocksFromResponse = responseMessage.getBlocks();
         logger.debug("CLIENT - Processing snap blocks response. Receiving from block {} to block {} Objective: {}.", blocksFromResponse.get(0).getNumber(), blocksFromResponse.get(blocksFromResponse.size() - 1).getNumber(), lastRequiredBlock);
         List<BlockDifficulty> difficultiesFromResponse = responseMessage.getDifficulties();
 
-        for (int i = 0; i < blocksFromResponse.size(); i++) {
-            state.addBlock(new ImmutablePair<>(blocksFromResponse.get(i), difficultiesFromResponse.get(i)));
+        if (!validateAndSaveBlocks(state, sender, blocksFromResponse, difficultiesFromResponse)) {
+            return;
         }
+
         long nextChunk = blocksFromResponse.get(0).getNumber();
         logger.debug("CLIENT - SnapBlock - nexChunk : {} - lastRequired {}, missing {}", nextChunk, lastRequiredBlock, nextChunk - lastRequiredBlock);
-        if (nextChunk > lastRequiredBlock) {
+
+        if (blocksVerified(state)) {
+            logger.info("CLIENT - Finished Snap blocks request sending.");
+
+            generateChunkRequestTasks(state);
+            startRequestingChunks(state);
+        } else if (nextChunk > lastRequiredBlock) {
             requestBlocksChunk(sender, nextChunk);
         } else {
             logger.info("CLIENT - Finished Snap blocks request sending.");
+
+            generateChunkRequestTasks(state);
+            startRequestingChunks(state);
+
+            requestNextBlockHeadersChunk(state, sender);
         }
```

Comment:
suggestion:

I see that the difference between the first if and the last else it's because we are requiring headers chunk if block is not verified. Maybe it could be simplified and avoid a bit of DRY:

```suggestion
        if (nextChunk > lastRequiredBlock) {
            requestBlocksChunk(sender, nextChunk);
        } else {
            logger.info("CLIENT - Finished Snap blocks request sending.");

            generateChunkRequestTasks(state);
            startRequestingChunks(state);
            if(!blocksVerified(state)) {
                requestNextBlockHeadersChunk(state, sender);
            }
        }
```